src/server/search_family.cc

```diff
@@ -0,0 +1,252 @@
+      out[SdsToSafeSv(k)] = SdsToSafeSv(v);
+    return out;
+  }
+
+ private:
+  StringMap* hset_;
+};
+
+variant<ListPackAccessor, StringMapAccessor> GetAccessor(const OpArgs& op_args,
+                                                         const PrimeValue& pv) {
+  if (pv.Encoding() == kEncodingListPack) {
+    auto ptr = reinterpret_cast<ListPackAccessor::LpPtr>(pv.RObjPtr());
+    return ListPackAccessor{ptr};
+  } else {
+    DCHECK_EQ(pv.Encoding(), kEncodingStrMap2);
+    auto* sm = container_utils::GetStringMap(pv, op_args.db_cntx);
+    return StringMapAccessor{sm};
+  }
+}
+
+void OpSearch(const OpArgs& op_args, vector<SerializedDocument>* shard_out,
+              search::AstExpr filter) {
+  auto& db_slice = op_args.shard->db_slice();
+  DCHECK(db_slice.IsDbValid(op_args.db_cntx.db_index));
+
+  auto [prime_table, expire_table] = db_slice.GetTables(op_args.db_cntx.db_index);
+
+  auto cb = [&](PrimeTable::iterator it) {
+    const PrimeValue& pv = it->second;
+    if (pv.ObjType() != OBJ_HASH)
+      return;
+
+    auto stored_accessor = GetAccessor(op_args, pv);
+    BaseAccessor* accessor =
+        visit([](auto& o) { return static_cast<BaseAccessor*>(&o); }, stored_accessor);
+
+    VLOG(0) << "Checking " << it->first.ToString() << " res "
+            << filter->Check(search::SearchInput{accessor});
+
+    if (bool res = filter->Check(search::SearchInput{accessor}); !res)
+      return;
+
+    VLOG(0) << "Its a match";
+    shard_out->push_back(accessor->Serialize());
+  };
+
+  PrimeTable::Cursor cursor;
+  do {
+    cursor = prime_table->Traverse(cursor, cb);
+  } while (cursor);
+}
+
+}  // namespace
+
+void SearchFamily::FtCreate(CmdArgList args, ConnectionContext* cntx) {
+  string_view idx = ArgS(args, 0);
+  string_view prefix = ArgS(args, 5);
+  {
+    CHECK_EQ(ArgS(args, 1), "ON");
+    CHECK_EQ(ArgS(args, 2), "HASH");
+    CHECK_EQ(ArgS(args, 3), "PREFIX");
+    CHECK_EQ(ArgS(args, 4), "1");
+  }
+  {
+    lock_guard lk{prefixes_mu_};
+    prefixes_[idx] = prefix;
+  }
+  (*cntx)->SendOk();
+}
+
+void SearchFamily::FtSearch(CmdArgList args, ConnectionContext* cntx) {
+  string_view idx = ArgS(args, 0);
+  string_view query = ArgS(args, 1);
+
+  // Find prefix for index.
+  string_view prefix;
+  {
+    lock_guard lk{prefixes_mu_};
+    auto it = prefixes_.find(idx);
+    if (it == prefixes_.end()) {
+      (*cntx)->SendError("Search index not found");
+      return;
+    }
+    prefix = it->second;
+  }
+
+  if (prefix.empty()) {
+    (*cntx)->SendError("Invalid index");
+    return;
+  }
+
+  search::AstExpr filter = search::ParseQuery(query);
+  if (!filter) {
+    (*cntx)->SendError("Invalid query");
+    return;
+  }
+
+  VLOG(0) << "Filter is " << filter->Debug();
+
+  vector<vector<SerializedDocument>> docs(shard_set->size());
+  cntx->transaction->ScheduleSingleHop([&](Transaction* t, EngineShard* shard) {
+    OpSearch(t->GetOpArgs(shard), &docs[shard->shard_id()], filter);
```

Comment:
We can have some contention here when pushing to the vectors... We have this pattern also in other places in the code as well, but this one is a bit worse because you're pushing to the vector in a loop during the search :)

Reply:
Do you mean that we push in a loop without a resize ahead? We don't have any estimation for the number of docs 🤷🏻 I don't think that's as issue for now

Reply:
No, sorry that I've been unclear. We're pushing to `docs[0]` and `docs[1]` from different threads, but since they're allocated continuously inside `docs` they're probably on the same cache line. And since each `push_back` changes the vector's size member, you're going to get contention between the threads.

Agree that's not the end of the world, just wanted to point it out :)

Reply:
Ah, I see, you mean cache padding. Agree, we need a general solution for this for all other places as well:
- a CachePadded<T> wrapper
- and possibly use an inlined vector for small thread counts

Transaction currently does all of this with its shard data, but its not re-usable.